packages/flutter/lib/src/material/carousel.dart

```diff
@@ -0,0 +1,744 @@
+// Copyright 2014 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:math' as math;
+
+import 'package:flutter/foundation.dart';
+import 'package:flutter/rendering.dart';
+import 'package:flutter/widgets.dart';
+
+import 'colors.dart';
+import 'ink_well.dart';
+import 'material.dart';
+import 'theme.dart';
+
+/// A Material Design carousel.
+///
+/// This is a scrollable list where the size of each item can change dynamically
+/// based on different layouts.
+///
+/// Material Design 3 introduces 4 layouts for [Carousel]:
+///  * Multi-browse: This layout shows at least one large, medium, and small
+/// carousel item at a time.
+///  * Uncontained: This layout show items that scroll to the edge of the container.
+///  * Hero: This layout shows at least one large and one small item at a time.
+///  * Full-screen: This layout shows one edge-to-edge large item at a time and
+/// scrolls vertically.
+///
+/// By default, [Carousel] has a uncontained layout. It shows like a [ListView]
```

Comment:
Do we follow-up with the other layouts (perhaps with factory constructors) in the future?

Reply:
Yes, another PR is created here: https://github.com/QuncCccccc/flutter/pull/2